### DISCUSSION TITLE
Start multiple instances from Python

### DISCUSSION THREAD

**Discussion by AlTech98** [mitmproxy/mitmproxy#5762] [Q&A]
I'm trying to start multiple mitmweb instances from python, running on different ports and with the wireguard mode.

Is there a way to do so?
I tried with multiple threads and with asyncio tasks but every time this error is returned:
```
WireGuard server failed to listen on *:45678 with Address already in use (os error 98)
```
Or other random errors

Code:
```python
async def start_mitmweb(mitmweb_port, wireguard_port):
	opts = options.Options()
	master = web.master.WebMaster(opts)
	parser = cmdline.mitmweb(opts)
	args = parser.parse_args([
		"--mode", f"wireguard@{wireguard_port}",
		"--web-port", f"{mitmweb_port}"
	])
	process_options(parser, opts, args)
	# master.addons.remove(master.addons.get("onboarding"))
	return await master.run()
```
With threads:
```python
def thread1():
	asyncio.run(start_mitmweb(8081, 45678))

def thread2():
	asyncio.run(start_mitmweb(8082, 45690))

t1 = threading.Thread(target=thread1)
t2 = threading.Thread(target=thread2)
t1.start()
t2.start()
```
Asyncio loop:
```python
loop = asyncio.new_event_loop()
loop.create_task(start_mitmweb(8082, 45690))
loop.create_task(start_mitmweb(8081, 45678))
loop.run_forever()
```

**Comment by mhils** [comment]
Doing this is not supported. There are two alternatives:

1. You spawn each mitmproxy instance as a subprocess.
2. You pass `--mode` multiple times to one instance.

**Comment by AlTech98** [comment]
Thanks for the answer,
Is there a way in mitmweb to toggle the view between the modes or to add a column with some text added by an Addon in order to tell from which mode a flow comes from?
